src/main/java/com/spikes2212/control/TalonPIDLoop.java

```diff
@@ -0,0 +1,162 @@
+        this(motor, kP, kI, kD, setpoint, tolerance, canMove, waitTime, ControlMode.Position);
+    }
+
+    public TalonPIDLoop(BaseMotorController motor, Supplier<Double> kP, Supplier<Double> kI, Supplier<Double> kD,
+                        Supplier<Double> setpoint, Supplier<Double> tolerance, Function<Double, Boolean> canMove,
+                        Supplier<Double> waitTime, ControlMode controlMode) {
+        this(motor, kP, kI, kD, setpoint, tolerance, canMove, waitTime, controlMode, 0);
+    }
+
+    public TalonPIDLoop(BaseMotorController motor, Supplier<Double> kP, Supplier<Double> kI, Supplier<Double> kD,
+                        Supplier<Double> setpoint, Supplier<Double> tolerance, Function<Double, Boolean> canMove,
+                        Supplier<Double> waitTime, ControlMode controlMode, int loop) {
+        this(motor, kP, kI, kD, setpoint, tolerance, canMove, waitTime, controlMode, loop, 30);
+    }
+
+    public TalonPIDLoop(BaseMotorController motor, Supplier<Double> kP, Supplier<Double> kI, Supplier<Double> kD,
+                        Supplier<Double> setpoint, Supplier<Double> tolerance, Function<Double, Boolean> canMove,
+                        Supplier<Double> waitTime, ControlMode controlMode, int loop, int timeout) {
+        this.motor = motor;
+        this.kP = kP;
+        this.kI = kI;
+        this.kD = kD;
+        this.setpoint = setpoint;
+        this.tolerance = tolerance;
+        this.canMove = canMove;
+        this.waitTime = waitTime;
+        this.controlMode = controlMode;
+        this.loop = loop;
+        this.timeout = timeout;
+        this.lastTimeNotOnTarget = Timer.getFPGATimestamp();
+    }
+
+    private void initialize() {
+        motor.configFactoryDefault();
+        motor.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, loop, timeout);
+
+        motor.configNominalOutputForward(0, timeout);
+        motor.configNominalOutputReverse(0, timeout);
+        motor.configPeakOutputForward(1.0, timeout);
+        motor.configPeakOutputReverse(-1.0, timeout);
+
+        motor.configAllowableClosedloopError(loop, 0, timeout);
+
+        motor.config_kP(loop, kP.get(), timeout);
+        motor.config_kI(loop, kI.get(), timeout);
+        motor.config_kD(loop, kD.get(), timeout);
+
+        motor.setSelectedSensorPosition(0, loop, timeout);
+    }
+
+    @Override
+    public void enable() {
+        initialize();
+    }
+
+    @Override
+    public void disable() {
+        motor.set(ControlMode.PercentOutput, 0);
+    }
+
+    @Override
+    public void update() {
+        motor.config_kP(loop, kP.get(), timeout);
+        motor.config_kI(loop, kI.get(), timeout);
+        motor.config_kD(loop, kD.get(), timeout);
+
+        motor.set(controlMode, setpoint.get());
+    }
+
+    @Override
+    public boolean onTarget() {
+        boolean currentlyOnTarget = Math.abs(setpoint.get() - motor.getSelectedSensorPosition(loop)) < tolerance.get();
```

Comment:
this wouldn't neccesserally use `getSelectedSensorPosition` please find a way to make this more generic